src/services/MatchReportService.ts

```diff
@@ -0,0 +1,119 @@
+import ActiveMatchRepository from "../repositories/ActiveMatchRepository";
+import { ActiveMatchTeams, PlayerInActiveMatch } from "../repositories/ActiveMatchRepository/types";
+import LeaderboardRepository from "../repositories/LeaderboardRepository";
+import { UpdatePlayerStatsInput } from "../repositories/LeaderboardRepository/types";
+import { Team } from "../types/common";
+import { waitForAllPromises } from "../utils";
+
+async function calculateProbabilityDecimal(playerInMatchId: string, reportedTeam: Team): Promise<number> {
+  let reportedWinner = 0;
+  let reportedLoser = 0;
+  const teams = await ActiveMatchRepository.getAllPlayersInActiveMatch(playerInMatchId);
+  const blueTeam = teams.blueTeam;
+  const orangeTeam = teams.orangeTeam;
+
+  const blueTeamMMR = blueTeam.reduce((ballchaser1, ballchaser2) => ballchaser1 + ballchaser2.mmr, 0);
+  const orangeTeamMMR = orangeTeam.reduce((ballchaser1, ballchaser2) => ballchaser1 + ballchaser2.mmr, 0);
+
+  if (reportedTeam === Team.Blue) {
+    reportedWinner = blueTeamMMR;
+    reportedLoser = orangeTeamMMR;
+  } else {
+    reportedWinner = orangeTeamMMR;
+    reportedLoser = blueTeamMMR;
+  }
+
+  const difference = (reportedLoser - reportedWinner) / 400;
+  const power = Math.pow(10, difference) + 1;
+  const probabilityDecimal = 1 / power;
+
+  return probabilityDecimal;
+}
+
+export async function calculateMMR(playerInMatchId: string, reportedTeam: Team): Promise<number> {
+  const probabilityDecimal = await calculateProbabilityDecimal(playerInMatchId, reportedTeam);
+
+  let mmr = (1 - probabilityDecimal) * 20;
+  mmr = Math.min(15, mmr);
+  mmr = Math.max(5, mmr);
+
+  return Math.round(mmr);
+}
+
+export async function calculateProbability(playerInMatchId: string, reportedTeam: Team): Promise<number> {
+  const probabilityDecimal = await calculateProbabilityDecimal(playerInMatchId, reportedTeam);
+  const probability = probabilityDecimal * 100;
+
+  return Math.round(probability);
+}
+
+export async function checkReport(team: Team, playerInMatchId: string): Promise<boolean> {
+  const teams = await ActiveMatchRepository.getAllPlayersInActiveMatch(playerInMatchId);
+  const reporter = [...teams.blueTeam, ...teams.orangeTeam].find((p) => {
+    return p.id === playerInMatchId;
+  });
+
+  const reportedPlayer = [...teams.blueTeam, ...teams.orangeTeam].find((p) => {
+    return p.reportedTeam !== null;
+  });
+
+  if (!reporter) return false;
```

Comment:
Would it make sense to include a check for `reportedPlayer` here as well? Then you won't need the `?` with `reportedPlayer` below here.

Reply:
`reportedPlayer` may or may not exist depending on if it's the first report. It's okay if it does not exist which is why I have the `?` for subsequent checks.  I don't want to stop the code if `reportedPlayer` does not exist. I will rename this to `previousReportedPlayer` so it can make more sense.

Reply:
I do want to check if `!reporter` because they need to be in the current active match.

Reply:
I am also renaming `team` to be `reportedTeam`

Reply:
Now, pseudocode for the checks for reporting or confirming follow this convention:
if the reporter doesn't exist -> stop
if the previous reporter exists - does their team = the reporter's team? && do their ids match? -> stop
*This is to make sure that if the same person reports their team again, Norm doesn't run all the code again.
else if the previous reporter exists - does their reported team = current reported team? || is the same person reporting again? -> reportMatch
*This is to make sure that if the previous reported team and the current reported team aren't the same, or if the same person is reporting the opposite team, then send the report
else we confirm the match since these checks require two people from opposite teams in the same active match to report the same team who can't be the same person reporting twice.

Reply:
> `reportedPlayer` may or may not exist depending on if it's the first report. It's okay if it does not exist which is why I have the `?` for subsequent checks. I don't want to stop the code if `reportedPlayer` does not exist. I will rename this to `previousReporter` so it can make more sense.

Gotcha, that makes sense.